src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs

```diff
@@ -193,6 +190,64 @@ bool IHttpBodyControlFeature.AllowSynchronousIO
             }
         }
 
+        PipeWriter IResponseBodyPipeFeature.ResponseBodyPipe
+        {
+            get
+            {
+                if (_userSetPipeWriter != null)
+                {
+                    return _userSetPipeWriter;
+                }
+
+                if (!object.ReferenceEquals(_cachedResponseBodyStream, ResponseBody))
+                {
+                    var responsePipeWriter = new StreamPipeWriter(ResponseBody);
+                    ResponsePipeWriter = responsePipeWriter;
+                    _cachedResponseBodyStream = ResponseBody;
+                    OnCompleted((rpw) =>
+                    {
+                        ((StreamPipeWriter)rpw).Dispose();
```

Comment:
I would rather we didn't use a public OnCompleted method for this because intermingling StreamPipeWriter.Dispose() with user code seems like a recipe for problems down the road. If you were to dispose the StreamPipeWriter here, you should at least set ResponseBodyPipe back to the previous value.

Reply:
+1

Reply:
Few things:
- Because we are creating the StreamPipeWriters/WriteOnlyPipeStream, we need to handle disposing them. That means if multiple wrappers are created, we need to keep a list of them to dispose in Reset.
- I wanted to keep the solution the same between the default ResponsePipeFeature and Kestrel's implementation, but as we know exactly when the object will be dispose in Kestrel, we should take advantage of it. 
- If we Dispose on Reset, I don't think we need to bother with setting the ResponseBodyPipe to the previous value.

Reply:
I have a prototype of having a list of disposable objects here. Tell me if you like the idea, and I'll clean it up if it works.